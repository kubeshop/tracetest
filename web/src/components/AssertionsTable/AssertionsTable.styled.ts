import {Badge} from 'antd';
import styled from 'styled-components';

export const AssertionsTableContainer = styled.div`
  margin-bottom: 32px;
`;

export const AssertionsTableHeader = styled.div`
  padding: 8px 0;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

export const AssertionsTableBadge = styled(Badge)`
  > sup {
    background-color: #f0f0f0;
    color: black;
    margin-left: 6px;
  }
`;
