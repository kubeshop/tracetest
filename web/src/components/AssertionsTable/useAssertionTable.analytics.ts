import useAnalytics, {Categories} from '../Analytics/useAnalytics';

enum Actions {
  EditAssertionButtonClick = 'edit-assertion-button-click',
}

type TAssertionTableAnalytics = {
  onEditAssertionButtonClick(assertionId: string): void;
};

const useAssertionTableAnalytics = (): TAssertionTableAnalytics => {
  const {event} = useAnalytics(Categories.SpanDetail);

  const onEditAssertionButtonClick = (assertionId: string) => {
    event(Actions.EditAssertionButtonClick, assertionId);
  };

  return {
    onEditAssertionButtonClick,
  };
};

export default useAssertionTableAnalytics;
