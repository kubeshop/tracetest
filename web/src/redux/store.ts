import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import TestAPI from 'redux/apis/Test.api';
import ResultListSlice from 'redux/slices/ResultList.slice';

export const store = configureStore({
  reducer: {
    [TestAPI.reducerPath]: TestAPI.reducer,
    resultList: ResultListSlice,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TestAPI.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
