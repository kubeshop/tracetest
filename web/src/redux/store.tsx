import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {createReduxHistoryContext} from 'redux-first-history';
import {createBrowserHistory} from 'history';
import TracetestAPI from 'redux/apis/Tracetest';

import RouterMiddleware from './Router.middleware';
import {middlewares, reducers} from './setup';

const {createReduxHistory, routerMiddleware, routerReducer} = createReduxHistoryContext({
  history: createBrowserHistory(),
});

TracetestAPI.create();

export const store = configureStore({
  reducer: {
    ...reducers,
    [TracetestAPI.instance.reducerPath]: TracetestAPI.instance.reducer,
    router: routerReducer,
  },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware()
      .prepend(RouterMiddleware.middleware)
      .concat(TracetestAPI.instance.middleware, ...middlewares, routerMiddleware),
});

export const history = createReduxHistory(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
