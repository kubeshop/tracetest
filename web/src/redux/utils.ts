import {PromiseWithKnownReason} from '@reduxjs/toolkit/dist/query/core/buildMiddleware/types';
import webSocketGateway, {IListenerFunction} from 'gateways/WebSocket.gateway';

export type {IListenerFunction} from 'gateways/WebSocket.gateway';

interface IInitWebSocketSubscription {
  listener: IListenerFunction;
  resource: string;
  waitToCleanSubscription: Promise<void>;
  waitToInitSubscription: PromiseWithKnownReason<any, any>;
}

/**
 * Implements RTK Query logic to receive streaming updates for persistent queries.
 * It allows a query to establish a WebSocket subscription, and apply updates to
 * the cached data when the information is received from the server.
 * For more information please take a look: https://redux-toolkit.js.org/rtk-query/usage/streaming-updates
 */
export async function initWebSocketSubscription({
  listener,
  resource,
  waitToCleanSubscription,
  waitToInitSubscription,
}: IInitWebSocketSubscription) {
  try {
    await waitToInitSubscription;
    webSocketGateway.subscribe(resource, listener);
  } catch {
    // no-op in case `waitToCleanSubscription` resolves before `waitToInitSubscription`,
    // in which case `waitToInitSubscription` will throw
  }
  await waitToCleanSubscription;
  webSocketGateway.unsubscribe(resource, '1234');
}
