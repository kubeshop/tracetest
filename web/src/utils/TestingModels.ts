import {HTTP_METHOD} from '../constants/Common.constants';
import {CompareOperator} from '../constants/Operator.constants';
import {SemanticGroupNames} from '../constants/SemanticGroupNames.constants';
import {LOCATION_NAME} from '../constants/Span.constants';
import {SpanAttributeType} from '../constants/SpanAttribute.constants';
import {TestState} from '../constants/TestRunResult.constants';
import {IAssertion, IAssertionResult, IItemSelector, ISpanAssertionResult} from '../types/Assertion.types';
import {ISpan} from '../types/Span.types';
import {ITest} from '../types/Test.types';
import {ITestRunResult} from '../types/TestRunResult.types';
import {ITrace} from '../types/Trace.types';

interface IProps {
  mouseEvent: MouseEvent;
  assertionId: string;
  spanAssertionResult: ISpanAssertionResult;
  trace: ITrace;
  span: ISpan;
  test: ITest;
  assertion: IAssertion;
  assertionResult: IAssertionResult;
  testRunResult: ITestRunResult;
  resultId: string;
  testId: string;
  spanId: string;
}

const resultId = '0932u0e3';
const testId = '2309ei30';
const spanId = '4938u43';

const test: ITest = {
  assertions: [],
  description: '',
  lastTestResult: undefined,
  name: '',
  serviceUnderTest: {
    id: '',
    request: {
      auth: undefined,
      body: '',
      certificate: undefined,
      headers: undefined,
      method: HTTP_METHOD.GET,
      proxy: undefined,
      url: '',
    },
  },
  testId,
};

const span: ISpan = {
  attributeList: [],
  attributes: {},
  duration: 0,
  endTimeUnixNano: '',
  kind: '',
  name: '',
  parentSpanId: '',
  signature: [
    {
      locationName: LOCATION_NAME.SPAN,
      propertyName: '',
      value: '',
      valueType: '',
    },
  ],
  spanId,
  startTimeUnixNano: '',
  status: {code: ''},
  traceId: '',
  type: SemanticGroupNames.Http,
};

let selector: IItemSelector = {
  locationName: LOCATION_NAME.SPAN,
  propertyName: '',
  value: '',
  valueType: '',
};

const assertionResult: IAssertionResult = {
  assertion: {
    assertionId: '',
    selectors: [selector],
    spanAssertions: [],
  },
  spanListAssertionResult: [
    {
      resultList: [
        {
          actualValue: 'New request',
          comparisonValue: '',
          hasPassed: false,
          locationName: LOCATION_NAME.SPAN,
          operator: CompareOperator.EQUALS,
          propertyName: '',
          spanAssertionId: '',
          spanId: '',
          valueType: SpanAttributeType.intValue,
        },
      ],
      span,
    },
  ],
};

const testRunResult: ITestRunResult = {
  executionTime: 0,
  failedAssertionCount: 0,
  passedAssertionCount: 0,
  totalAssertionCount: 0,

  assertionResult: [{assertionId: '', spanAssertionResults: []}],
  assertionResultState: false,
  completedAt: '',
  createdAt: new Date().toISOString().toString(),
  response: undefined,
  resultId: '',
  spanId: '',
  state: TestState.CREATED,
  testId: '',
  trace: {
    description: '',
    resourceSpans: undefined,
    spans: [span],
  },
  traceId: '',
};

const spanAssertionResult: ISpanAssertionResult = {
  actualValue: 'New request',
  comparisonValue: '',
  hasPassed: false,
  locationName: LOCATION_NAME.INSTRUMENTATION_LIBRARY,
  operator: CompareOperator.EQUALS,
  propertyName: '',
  spanAssertionId: '',
  spanId: '',
  valueType: SpanAttributeType.doubleValue,
};

const assertionId = '234234';

const assertion: IAssertion = {
  assertionId,
  selectors: [
    {
      locationName: LOCATION_NAME.INSTRUMENTATION_LIBRARY,
      propertyName: '',
      value: '',
      valueType: '',
    },
  ],
  spanAssertions: [],
};
export const TestingModels: IProps = {
  mouseEvent: new MouseEvent('click', {
    bubbles: true,
    cancelable: true,
  }),
  assertionId,
  assertion,
  spanAssertionResult,
  assertionResult,
  test,
  testRunResult,
  span,
  resultId,
  testId,
  spanId,
  trace: {
    description: '',
    spans: [span],
  },
};

class NewTestingModel {
  test: ITest = test;
  assertionResult: IAssertionResult = assertionResult;
  testRunResult: ITestRunResult = testRunResult;
  span: ISpan = span;
  resultId: string = resultId;
  testId: string = testId;
  spanId: string = spanId;
  trace: ITrace = {
    description: '',
    spans: [span],
  };
}

export const TestingObj = new NewTestingModel();
