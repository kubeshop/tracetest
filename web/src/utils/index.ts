import {COMPARE_OPERATOR} from 'types';

const flattenAttributesSelector = () =>
  `resourceSpans[].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0]},resource.attributes[].{key:key,value: value.*|[0]}]|[][]`;

export const filterBySpanId = (spanId: string = '') =>
  `resourceSpans[?instrumentationLibrarySpans[?spans[?starts_with(spanId,'${spanId}')]]] | [].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0],type:'span'},resource.attributes[].{key:key,value: value.*|[0],type:'resource'}]|[][]`;

export const getOperator = (op: COMPARE_OPERATOR) => {
  switch (op) {
    case COMPARE_OPERATOR.EQUALS:
      return 'eq';
    case COMPARE_OPERATOR.NOTEQUALS:
      return 'ne';
    case COMPARE_OPERATOR.GREATERTHAN:
      return 'gt';
    case COMPARE_OPERATOR.LESSTHAN:
      return 'lt';
    case COMPARE_OPERATOR.GREATOREQUALS:
      return 'gte';
    case COMPARE_OPERATOR.LESSOREQUAL:
      return 'lte';
    default:
      return 'eq';
  }
};
