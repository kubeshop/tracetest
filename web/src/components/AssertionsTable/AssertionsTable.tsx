import {Button, Table, Typography} from 'antd';
import {useMemo} from 'react';
import {Assertion, SpanAssertionResult} from 'types';
import {getOperator} from 'utils';
import CustomTable from '../CustomTable';
import * as S from './AssertionsTable.styled';

interface IProps {
  assertionResults: SpanAssertionResult[];
  assertion: Assertion;
  sort: number;
}

type TParsedAssertion = {
  key: string;
  property: string;
  comparison: string;
  value: string;
  actualValue: string;
  hasPassed: boolean;
};

const AssertionsResultTable = ({assertionResults, assertion: {selectors = []}, sort}: IProps) => {
  const parsedAssertionList = useMemo<Array<TParsedAssertion>>(
    () =>
      assertionResults.map(({propertyName, comparisonValue, operator, actualValue, hasPassed}) => ({
        key: propertyName,
        property: propertyName,
        comparison: operator,
        value: comparisonValue,
        actualValue,
        hasPassed,
      })),
    [assertionResults]
  );

  return (
    <S.AssertionsTableContainer>
      <S.AssertionsTableHeader>
        <Typography.Title level={3} style={{margin: 0}}>
          Assertion #{sort}
          {selectors.map(({value, propertyName}) => (
            <S.AssertionsTableBadge count={value} key={propertyName} />
          ))}
        </Typography.Title>
        <Button type="link">Edit</Button>
      </S.AssertionsTableHeader>
      <CustomTable
        size="small"
        pagination={{hideOnSinglePage: true}}
        dataSource={parsedAssertionList}
        bordered
        tableLayout="fixed"
      >
        <Table.Column title="Property" dataIndex="property" key="property" ellipsis width="60%" />
        <Table.Column title="Comparison" dataIndex="comparison" key="comparison" render={value => getOperator(value)} />
        <Table.Column title="Value" dataIndex="value" key="value" />
        <Table.Column
          title="Actual"
          dataIndex="actualValue"
          key="actualValue"
          render={(value, record: TParsedAssertion) => (
            <Typography.Text strong type={record.hasPassed ? 'success' : 'danger'}>
              {value}
            </Typography.Text>
          )}
        />
      </CustomTable>
    </S.AssertionsTableContainer>
  );
};

export default AssertionsResultTable;
