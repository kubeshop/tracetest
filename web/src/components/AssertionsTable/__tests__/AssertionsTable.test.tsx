import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {store} from '../../../redux/store';
import {TestingModels} from '../../../utils/TestingModels';
import AssertionsTable from '../index';

test('AssetionsTable', () => {
  const result = render(
    <Provider store={store}>
      <AssertionsTable
        assertionResults={[TestingModels.spanAssertionResult]}
        sort={1}
        assertion={TestingModels.assertion}
        span={TestingModels.span}
        resultId={TestingModels.resultId}
        testId={TestingModels.testId}
      />
    </Provider>
  );
  expect(result.container).toMatchSnapshot();
});
