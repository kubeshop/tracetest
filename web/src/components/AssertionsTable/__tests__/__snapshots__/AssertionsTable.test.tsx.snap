// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetionsTable 1`] = `
<div>
  <div
    class="sc-gsDKAQ hZxeIA"
    data-cy="assertion-table"
  >
    <div
      class="sc-dkPtRN diGMvB"
    >
      <h5
        class="ant-typography"
        style="margin: 0px;"
      >
        Assertion #
        1
        <span
          class="ant-badge ant-badge-not-a-wrapper sc-hKwDye wxfDf"
        />
      </h5>
      <button
        class="ant-btn ant-btn-link"
        data-cy="edit-assertion-button"
        type="button"
      >
        <span>
          Edit
        </span>
      </button>
    </div>
    <div
      class="ant-table-wrapper sc-bdvvtL kLJHSf"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-small ant-table-bordered ant-table-layout-fixed"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <colgroup>
                    <col
                      style="width: 50%;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title="Property"
                      >
                        Property
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Comparison
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Value
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Actual
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key=""
                    >
                      <td
                        class="ant-table-cell ant-table-cell-ellipsis"
                        title=""
                      >
                        <span
                          data-cy="assertion-check-property"
                        />
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        equals
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        <span
                          class="ant-typography ant-typography-danger"
                        >
                          <strong>
                            New request
                          </strong>
                        </span>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
